client/consensus/src/collators/lookahead.rs

```diff
@@ -74,8 +85,255 @@ use {
+        + HeaderBackend<OBlock>
+        + BlockBackend<OBlock>
+        + BlockIdTo<OBlock>
+        + 'static,
+    OClient::Api: TaggedTransactionQueue<OBlock>
+        + XCMCoreBuyerApi<OBlock, <<OBlock as BlockT>::Header as HeaderT>::Number, ParaId, NimbusId>,
+{
+    // We do nothing if this is not a parathread
+    if aux_data.slot_freq.is_none() {
+        return Err(BuyCoreError::NotAParathread);
+    }
+
+    let orchestrator_best_hash = orchestrator_client.info().best_hash;
+    let orchestrator_runtime_api = orchestrator_client.runtime_api();
+
+    let buy_core_slot_drift = orchestrator_client
+        .runtime_api()
+        .get_buy_core_slot_drift(orchestrator_best_hash)?;
+
+    // Convert drift in terms of container chain slots for parity between client side calculation and
+    // orchestrator runtime calculation
+    // We tried to made this calculation as generic as possible so that it can handle
+    // arbitrary slot timings as well and won't assume anything.
+    // Formula is: (Slot_drift_in_orchestrator_slot * orchestrator_slot_duration) / container_slot_duration
+    let buy_core_container_slot_drift = buy_core_slot_drift
+        .checked_mul(orchestrator_slot_duration.as_millis())
+        .and_then(|intermediate_result| {
+            intermediate_result.checked_div(container_slot_duration.as_millis())
+        })
+        .ok_or(BuyCoreError::SlotDriftConversionOverflow)?;
+
+    let current_container_slot = inherent_providers.slot();
+
+    let slot_claim = tanssi_claim_slot::<P, Block>(
+        aux_data,
+        &parent_header,
+        current_container_slot,
+        ClaimMode::ParathreadCoreBuying {
+            drift_permitted: buy_core_container_slot_drift.into(),
+        },
+        keystore,
+    )
+    .ok_or(BuyCoreError::UnableToClaimSlot)?;
+
+    let pubkey = slot_claim.author_pub;
+
+    orchestrator_runtime_api.is_core_buying_allowed(orchestrator_best_hash, para_id)??;
+
+    let nonce =
+        orchestrator_runtime_api.get_buy_core_signature_nonce(orchestrator_best_hash, para_id)?;
+
+    let collator_buy_core_proof =
+        BuyCoreCollatorProof::new_with_keystore(nonce, para_id, pubkey, keystore)?
+            .ok_or(BuyCoreError::UnableToFindKeyForSigning)?;
+
+    let extrinsic = orchestrator_runtime_api.create_buy_core_unsigned_extrinsic(
+        orchestrator_best_hash,
+        para_id,
+        collator_buy_core_proof,
+    )?;
+
+    let mut tx_stream = orchestrator_tx_pool
+        .submit_and_watch(orchestrator_best_hash, TransactionSource::Local, *extrinsic)
+        .await
+        .map_err(BuyCoreError::TxSubmissionError)?
+        .fuse();
+
+    let current_time = tokio::time::Instant::now();
+    // Wait at max for 3 orchestrator slots
+    let wait_till = current_time + (orchestrator_slot_duration.as_duration() * 3);
+    loop {
+        select! {
+            _ = tokio::time::sleep_until(wait_till) => return Err(BuyCoreError::TxInclusionError),
```

Comment:
Why no just `tokio::time::sleep(timeout)`? With `timeout = orchestrator_slot_duration.as_duration() * 3` defined above.

Reply:
Problem is I want to only wait for 3 slot, no matter how many times the status of tx changes before it is either in block or finalized. 

If I do `tokio::time::sleep(timeout)` it will wait everytime the status changes for 3 slots and in a scenario where tx stays in a pool and does not make it to block, it will wait far too long.

Reply:
Oh, I see, because of the outer `loop`. Can you add a comment about this?

Reply:
sure